File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private void solve(int n, List<String> res, int open, int close, String s)
    {
        if(open==n && close==n)
        {
            res.add(s);
            return;
        }
        if(close>open || open>n)
        {
            return;
        }
        solve(n,res,open+1,close,s+"(");
        solve(n,res,open,close+1,s+")");
    }

    public List<String> generateParenthesis(int n) {
        List<String> res= new ArrayList<>();
        solve(n, res, 0, 0, "");
        return res;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def solve(self, n: int, res: List[str], open: int, close: int, s: str):
        if(open==n and close==n):
            res.append(s)
            return
        if(close>open or open>n):
            return
        self.solve(n,res,open+1,close,s+"(")
        self.solve(n,res,open,close+1,s+")")
        
    def generateParenthesis(self, n: int) -> List[str]:
        res=[]
        self.solve(n, res, 0, 0, "")
        return res